modules/sdk/src/main/scala/org/tessellation/sdk/http/p2p/middlewares/PeerAuthMiddleware.scala

```diff
@@ -0,0 +1,216 @@
+package org.tessellation.sdk.http.p2p.middleware
+
+import java.security.{PrivateKey, PublicKey}
+
+import cats.data.{Kleisli, OptionT}
+import cats.effect.{Async, Ref, Resource}
+import cats.syntax.applicative._
+import cats.syntax.applicativeError._
+import cats.syntax.flatMap._
+import cats.syntax.functor._
+import cats.syntax.option._
+
+import org.tessellation.schema.cluster.TokenValid
+import org.tessellation.schema.peer.PeerId
+import org.tessellation.sdk.domain.cluster.services.Session
+import org.tessellation.sdk.domain.cluster.storage.SessionStorage
+import org.tessellation.sdk.http.p2p.headers.{`X-Id`, `X-Session-Token`}
+import org.tessellation.security.SecurityProvider
+import org.tessellation.security.signature.Signing
+
+import com.comcast.ip4s.Host
+import fs2.{Chunk, Stream}
+import org.http4s.Status.Unauthorized
+import org.http4s._
+import org.http4s.client.Client
+import org.typelevel.ci._
+import pl.abankowski.httpsigner.http4s._
+import pl.abankowski.httpsigner.signature.generic.{GenericGenerator, GenericVerifier}
+import pl.abankowski.httpsigner.signature.{Generator, Verifier}
+import pl.abankowski.httpsigner.{HttpCryptoConfig, SignatureValid}
+
+object PeerAuthMiddleware {
+
+  private def unauthorized[F[_]] = Response[F](status = Unauthorized)
+
+  private def getOwnTokenHeader[F[_]: Async](sessionStorage: SessionStorage[F]): F[Option[`X-Session-Token`]] =
+    sessionStorage.getToken.map(_.map(t => `X-Session-Token`(t)))
+
+  private def getHost[F[_]](req: Request[F]): Option[Host] =
+    req.uri.authority.map(_.host.value).flatMap(Host.fromString)
+
+  private def getRemoteHost[F[_]](req: Request[F]): Option[Host] =
+    req.remoteAddr
+      .flatMap(_.asIpv4)
+      .map(_.toString)
+      .flatMap(Host.fromString)
+
+  def responseSignerMiddleware[F[_]: Async: SecurityProvider](
+    privateKey: PrivateKey,
+    sessionStorage: SessionStorage[F]
+  )(http: HttpRoutes[F]): HttpRoutes[F] =
+    Kleisli { req: Request[F] =>
+      for {
+        res <- http(req)
+        headerToken <- getOwnTokenHeader(sessionStorage).attemptT.toOption
+        newHeaders = headerToken.fold(res.headers)(h => res.headers.put(h))
+        resWithHeader = res.copy(headers = newHeaders)
+        signedResponse <- new Http4sResponseSigner[F](getGenerator(privateKey), new TessellationHttpCryptoConfig {})
```

Comment:
I wonder do we need to create them every time we are singing something?

Reply:
I'm not sure but most likely we did it like that due to concurrency issues